File: apps/web/src/lib/youtube.ts
/**
 * YouTube URL validation and extraction utilities
 */

export interface YouTubeVideoInfo {
  videoId: string;
  url: string;
  isValid: boolean;
  error?: string;
}

/**
 * Extracts YouTube video ID from various URL formats
 * Supports:
 * - https://www.youtube.com/watch?v=VIDEO_ID
 * - https://youtu.be/VIDEO_ID
 * - https://www.youtube.com/embed/VIDEO_ID
 * - https://m.youtube.com/watch?v=VIDEO_ID
 */
export function extractVideoId(url: string): string | null {
  if (!url) return null;

  // Remove whitespace
  url = url.trim();

  // Standard watch URL
  const watchMatch = url.match(/[?&]v=([^&]+)/);
  if (watchMatch) return watchMatch[1];

  // Short URL
  const shortMatch = url.match(/youtu\.be\/([^?]+)/);
  if (shortMatch) return shortMatch[1];

  // Embed URL
  const embedMatch = url.match(/youtube\.com\/embed\/([^?]+)/);
  if (embedMatch) return embedMatch[1];

  // If it's just the video ID
  if (/^[a-zA-Z0-9_-]{11}$/.test(url)) return url;

  return null;
}

/**
 * Validates YouTube URL and extracts video information
 */
export function validateYouTubeUrl(url: string): YouTubeVideoInfo {
  const videoId = extractVideoId(url);

  if (!videoId) {
    return {
      videoId: '',
      url,
      isValid: false,
      error: 'Invalid YouTube URL. Please provide a valid YouTube video link.',
    };
  }

  // Validate video ID format (11 characters, alphanumeric plus - and _)
  if (!/^[a-zA-Z0-9_-]{11}$/.test(videoId)) {
    return {
      videoId,
      url,
      isValid: false,
      error: 'Invalid video ID format.',
    };
  }

  return {
    videoId,
    url: `https://www.youtube.com/watch?v=${videoId}`,
    isValid: true,
  };
}

/**
 * Generates YouTube thumbnail URL
 */
export function getThumbnailUrl(videoId: string, quality: 'default' | 'hq' | 'maxres' = 'hq'): string {
  const qualityMap = {
    default: 'default',
    hq: 'hqdefault',
    maxres: 'maxresdefault',
  };

  return `https://img.youtube.com/vi/${videoId}/${qualityMap[quality]}.jpg`;
}

/**
 * Generates YouTube embed URL
 */
export function getEmbedUrl(videoId: string, options?: {
  autoplay?: boolean;
  start?: number;
  end?: number;
}): string {
  const params = new URLSearchParams();

  if (options?.autoplay) params.set('autoplay', '1');
  if (options?.start) params.set('start', options.start.toString());
  if (options?.end) params.set('end', options.end.toString());

  const queryString = params.toString();
  return `https://www.youtube.com/embed/${videoId}${queryString ? '?' + queryString : ''}`;
}

/**
 * Converts timestamp string (HH:MM:SS or MM:SS) to seconds
 */
export function timestampToSeconds(timestamp: string): number {
  const parts = timestamp.split(':').map(Number);

  if (parts.length === 2) {
    // MM:SS
    return parts[0] * 60 + parts[1];
  } else if (parts.length === 3) {
    // HH:MM:SS
    return parts[0] * 3600 + parts[1] * 60 + parts[2];
  }

  return 0;
}

/**
 * Converts seconds to timestamp string (HH:MM:SS or MM:SS)
 */
export function secondsToTimestamp(seconds: number): string {
  const hours = Math.floor(seconds / 3600);
  const minutes = Math.floor((seconds % 3600) / 60);
  const secs = Math.floor(seconds % 60);

  if (hours > 0) {
    return `${hours}:${minutes.toString().padStart(2, '0')}:${secs.toString().padStart(2, '0')}`;
  }

  return `${minutes}:${secs.toString().padStart(2, '0')}`;
}

/**
 * Generates YouTube URL with timestamp
 */
export function getTimestampUrl(videoId: string, timestamp: string | number): string {
  const seconds = typeof timestamp === 'string' ? timestampToSeconds(timestamp) : timestamp;
  return `https://www.youtube.com/watch?v=${videoId}&t=${seconds}s`;
}

File: apps/backends/python-fastapi/app/tools/youtube.py
"""
YouTube utility functions
"""
import re
from typing import Optional
from urllib.parse import urlparse, parse_qs


def extract_video_id(url: str) -> Optional[str]:
    """
    Extract YouTube video ID from various URL formats

    Supported formats:
    - https://www.youtube.com/watch?v=VIDEO_ID
    - https://youtu.be/VIDEO_ID
    - https://www.youtube.com/embed/VIDEO_ID
    - https://www.youtube.com/v/VIDEO_ID

    Args:
        url: YouTube video URL

    Returns:
        Video ID or None if invalid
    """
    if not url:
        return None

    # Pattern for video ID (11 characters, alphanumeric + dash/underscore)
    video_id_pattern = r'[a-zA-Z0-9_-]{11}'

    # Try different URL formats
    patterns = [
        r'(?:youtube\.com\/watch\?v=)([a-zA-Z0-9_-]{11})',
        r'(?:youtu\.be\/)([a-zA-Z0-9_-]{11})',
        r'(?:youtube\.com\/embed\/)([a-zA-Z0-9_-]{11})',
        r'(?:youtube\.com\/v\/)([a-zA-Z0-9_-]{11})',
    ]

    for pattern in patterns:
        match = re.search(pattern, url)
        if match:
            return match.group(1)

    # Try parsing as URL with query parameters
    try:
        parsed = urlparse(url)
        if 'youtube.com' in parsed.netloc:
            query_params = parse_qs(parsed.query)
            if 'v' in query_params:
                video_id = query_params['v'][0]
                if re.match(video_id_pattern, video_id):
                    return video_id
    except:
        pass

    return None


def is_valid_youtube_url(url: str) -> bool:
    """
    Check if URL is a valid YouTube video URL

    Args:
        url: URL to validate

    Returns:
        True if valid YouTube URL
    """
    return extract_video_id(url) is not None


def get_thumbnail_url(video_id: str, quality: str = "maxresdefault") -> str:
    """
    Get YouTube thumbnail URL for a video

    Args:
        video_id: YouTube video ID
        quality: Thumbnail quality (maxresdefault, hqdefault, mqdefault, sddefault, default)

    Returns:
        Thumbnail URL
    """
    return f"https://img.youtube.com/vi/{video_id}/{quality}.jpg"


def format_duration(seconds: int) -> str:
    """
    Format duration in seconds to HH:MM:SS or MM:SS

    Args:
        seconds: Duration in seconds

    Returns:
        Formatted duration string
    """
    hours = seconds // 3600
    minutes = (seconds % 3600) // 60
    secs = seconds % 60

    if hours > 0:
        return f"{hours:02d}:{minutes:02d}:{secs:02d}"
    else:
        return f"{minutes:02d}:{secs:02d}"


def parse_duration(duration_str: str) -> int:
    """
    Parse duration string to seconds

    Args:
        duration_str: Duration string (HH:MM:SS or MM:SS)

    Returns:
        Duration in seconds
    """
    parts = duration_str.split(':')

    if len(parts) == 3:
        # HH:MM:SS
        return int(parts[0]) * 3600 + int(parts[1]) * 60 + int(parts[2])
    elif len(parts) == 2:
        # MM:SS
        return int(parts[0]) * 60 + int(parts[1])
    else:
        return 0
